homeassistant/components/moon/__init__.py

```diff
@@ -14,3 +31,97 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a config entry."""
     return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+def get_moon_phase(date: datetime.date) -> str:
+    """Get moon phase."""
+    state = moon.phase(date)
+
+    if state < 0.5 or state > 27.5:
+        return STATE_NEW_MOON
+    if state < 6.5:
+        return STATE_WAXING_CRESCENT
+    if state < 7.5:
+        return STATE_FIRST_QUARTER
+    if state < 13.5:
+        return STATE_WAXING_GIBBOUS
+    if state < 14.5:
+        return STATE_FULL_MOON
+    if state < 20.5:
+        return STATE_WANING_GIBBOUS
+    if state < 21.5:
+        return STATE_LAST_QUARTER
+    return STATE_WANING_CRESCENT
+
+
+def get_moon_phases(
+    start_date: datetime.date, end_date: datetime.date | None = None
+) -> list[dict[str, Any]]:
+    """Get moon phases for a range of dates."""
+
+    def generate_daterange(
+        start_date: datetime.date, end_date: datetime.date | None = None
+    ) -> Iterable[datetime.date]:
+        """Generate a range of dates."""
+        n_days = range(1)
+        if end_date is not None and end_date > start_date:
+            n_days = range(int((end_date - start_date).days))
+        for n in n_days:
+            yield start_date + datetime.timedelta(days=n)
+
+    def start_date_moon_phase(
+        current_date: datetime.date, current_phase: str
+    ) -> datetime.date:
+        """Get end date moon phase."""
+        start_moon_phase: datetime.date | None = None
+        for date_phase in generate_daterange(
+            current_date - datetime.timedelta(days=8), current_date
+        ):
+            if (
+                moon_phase := get_moon_phase(date_phase)
+            ) is not None and moon_phase == current_phase:
+                start_moon_phase = date_phase
+                break
+        if start_moon_phase is None:
+            return current_date
+        return start_moon_phase
+
+    def end_date_moon_phase(
+        current_date: datetime.date, current_phase: str
+    ) -> datetime.date:
+        """Get end date moon phase."""
+        date_moon_phase: datetime.date | None = None
+        for date_phase in generate_daterange(
+            current_date, current_date + datetime.timedelta(days=28)
+        ):
+            if (
+                moon_phase := get_moon_phase(date_phase)
+            ) is not None and moon_phase != current_phase:
+                date_moon_phase = date_phase
+                break
+        if date_moon_phase is None:
+            return current_date
+        return date_moon_phase
+
+    moon_phases = []
```

Comment:
I think using `itertools.groupby` could help simplify this quite a bit giving us an easy way to convert a sequence of values into a grouping of all the same values with the same key. The complexity here is doing some searches and loops to find values with the same range, but I think it would be easier to make a flat list of all the dates and phases, then do the grouping and filtering after.

```
def get_moon_phases(start_date: datetime.date, end_date: datetime.date) -> Generator[tuple[str, datetime.date], None, None]:
  day = start_date
  while day < end_date:
    yield (get_moon_phase(day), day)
    day += datetime.timedelta(days=1)

@dataclass
class MoonPhaseRange:
  moon_phase: str
  start_date: datetime.date
  end_date: datetime.date

def make_phase_ranges(moon_phases: Iterable[tuple[str, datetime.date]]) -> None:
  grouped = itertools.groupby(moon_phases, key=lambda x: x[0])
  for phase, values in grouped:
    values_list = list(values)
    start_date = values_list[0][1]
    end_date = values_list[-1][1]
    yield MoonPhaseRange(phase, start_date, end_date)

phases = get_moon_phases(START, END)
phase_ranges = make_phase_ranges(phases)

print(list(phase_ranges))
```

Then we can get output like this:
```
[MoonPhaseRange(moon_phase='waning_gibbous', start_date=datetime.date(2024, 1, 1), end_date=datetime.date(2024, 1, 3)), MoonPhaseRange(moon_phase='last_quarter', start_date=datetime.date(2024, 1, 4), end_date=datetime.date(2024, 1, 4)), MoonPhaseRange(moon_phase='waning_crescent', start_date=datetime.date(2024, 1, 5), end_date=datetime.date(2024, 1, 10)), MoonPhaseRange(moon_phase='new_moon', start_date=datetime.date(2024, 1, 11), end_date=datetime.date(2024, 1, 11)), MoonPhaseRange(moon_phase='waxing_crescent', start_date=datetime.date(2024, 1, 12), end_date=datetime.date(2024, 1, 17)), MoonPhaseRange(moon_phase='first_quarter', start_date=datetime.date(2024, 1, 18), end_date=datetime.date(2024, 1, 18)), MoonPhaseRange(moon_phase='waxing_gibbous', start_date=datetime.date(2024, 1, 19), end_date=datetime.date(2024, 1, 24)), MoonPhaseRange(moon_phase='full_moon', start_date=datetime.date(2024, 1, 25), end_date=datetime.date(2024, 1, 25)), MoonPhaseRange(moon_phase='waning_gibbous', start_date=datetime.date(2024, 1, 26), end_date=datetime.date(2024, 1, 30))]
```

So if you pad that with a broader date range then filter just to ranges that match the actual start/end.